In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


import warnings
warnings.filterwarnings('ignore')

## Loading the data

In [2]:
d2 = pd.read_csv(r"E:\Learning\TEAI Cup\Data\Main Data\5days_of_camera_data.csv", sep=';', encoding = 'latin1')

In [4]:
d2

device_name                 machine_definition  \
0        de512stmp223  spark-machine-otto-vision-dynamic   
1        de512stmp223  spark-machine-otto-vision-dynamic   
2        de512stmp226  spark-machine-otto-vision-dynamic   
3        de512stmp223  spark-machine-otto-vision-dynamic   
4        de512stmp226  spark-machine-otto-vision-dynamic   
...               ...                                ...   
2408391  de512stmp268  spark-machine-otto-vision-dynamic   
2408392  de512stmp268  spark-machine-otto-vision-dynamic   
2408393  de512stmp268  spark-machine-otto-vision-dynamic   
2408394  de512stmp268  spark-machine-otto-vision-dynamic   
2408395  de512stmp268  spark-machine-otto-vision-dynamic   

                    variable_name  \
0                       teil1GapX   
1                       teil1GapY   
2        iTeil2Rastfederdifferenz   
3          iT1RastfederDeformiert   
4              prfungTeil2BodySpk   
...                           ...   
2408391        spur2IKastenHinten   
2408392                 spur2Wire   
2408393                  spur2Iso   
2408394                  spur2Gap   
2408395     spur2DeformationFeder   

                                                     value  \
0        0.498,0.496,0.496,0.498,0.496,0.493,0.492,0.49...   
1        0.497,0.489,0.491,0.484,0.487,0.483,0.489,0.49...   
2        0.0107,0.0152,0.0074,0.0123,0.0007,0.0128,0.01...   
3        59.543,62.944,58.544,61.577,62.455,58.252,58.9...   
4        1.8878,1.8835,1.8934,1.8954,1.8851,1.8794,1.88...   
...                                                    ...   
2408391  0.979,0.981,0.986,0.976,0.982,0.983,0.983,0.97...   
2408392  1.713,1.712,1.713,1.713,1.714,1.714,1.715,1.71...   
2408393  1.936,1.939,1.936,1.932,1.936,1.933,1.931,1.93...   
2408394  0.194,0.192,0.184,0.194,0.189,0.2,0.203,0.207,...   
2408395  48.085,46.022,44.714,45.955,49.285,50.337,48.6...   

                   timestamp  \
0        2025-12-10 04:16:46   
1        2025-12-10 04:16:46   
2        2025-12-09 00:17:18   
3        2025-12-10 04:16:46   
4        2025-12-09 00:17:18   
...                      ...   
2408391  2025-12-06 08:16:17   
2408392  2025-12-06 08:16:17   
2408393  2025-12-06 08:16:17   
2408394  2025-12-06 08:16:17   
2408395  2025-12-06 08:16:17   

                                        variable_attribute  
0        lowerLimit: -0.06499, upperLimit: 0.00499, nom...  
1        lowerLimit: -0.06499, upperLimit: 0.00499, nom...  
2        lowerLimit: -0.08, upperLimit: 0.08, nominalVa...  
3        lowerLimit: -20, upperLimit: 22, nominalValue:...  
4        lowerLimit: 0, upperLimit: 0.35, nominalValue:...  
...                                                    ...  
2408391  lowerLimit: 0, upperLimit: 0.11, nominalValue:...  
2408392  lowerLimit: -0.1, upperLimit: 0.1, nominalValu...  
2408393  lowerLimit: -0.1, upperLimit: 0.1, nominalValu...  
2408394  lowerLimit: -0.034, upperLimit: 0.03, nominalV...  
2408395  lowerLimit: -25, upperLimit: 25, nominalValue:...  

[2408396 rows x 6 columns]

## Converting string time to pd. timstamp

In [5]:
d2.timestamp = pd.to_datetime(d2.timestamp)

In [6]:
d2.head()

device_name                 machine_definition             variable_name  \
0  de512stmp223  spark-machine-otto-vision-dynamic                 teil1GapX   
1  de512stmp223  spark-machine-otto-vision-dynamic                 teil1GapY   
2  de512stmp226  spark-machine-otto-vision-dynamic  iTeil2Rastfederdifferenz   
3  de512stmp223  spark-machine-otto-vision-dynamic    iT1RastfederDeformiert   
4  de512stmp226  spark-machine-otto-vision-dynamic        prfungTeil2BodySpk   

                                               value           timestamp  \
0  0.498,0.496,0.496,0.498,0.496,0.493,0.492,0.49... 2025-12-10 04:16:46   
1  0.497,0.489,0.491,0.484,0.487,0.483,0.489,0.49... 2025-12-10 04:16:46   
2  0.0107,0.0152,0.0074,0.0123,0.0007,0.0128,0.01... 2025-12-09 00:17:18   
3  59.543,62.944,58.544,61.577,62.455,58.252,58.9... 2025-12-10 04:16:46   
4  1.8878,1.8835,1.8934,1.8954,1.8851,1.8794,1.88... 2025-12-09 00:17:18   

                                  variable_attribute  
0  lowerLimit: -0.06499, upperLimit: 0.00499, nom...  
1  lowerLimit: -0.06499, upperLimit: 0.00499, nom...  
2  lowerLimit: -0.08, upperLimit: 0.08, nominalVa...  
3  lowerLimit: -20, upperLimit: 22, nominalValue:...  
4  lowerLimit: 0, upperLimit: 0.35, nominalValue:...

In [7]:
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2408396 entries, 0 to 2408395
Data columns (total 6 columns):
 #   Column              Dtype         
---  ------              -----         
 0   device_name         object        
 1   machine_definition  object        
 2   variable_name       object        
 3   value               object        
 4   timestamp           datetime64[ns]
 5   variable_attribute  object        
dtypes: datetime64[ns](1), object(5)
memory usage: 110.2+ MB


In [8]:
d2.isna().sum()

device_name           0
machine_definition    0
variable_name         0
value                 0
timestamp             0
variable_attribute    0
dtype: int64

In [9]:
d2.sort_values(by='timestamp', ascending=True, inplace=True)

In [10]:
d2.variable_attribute.sample().tolist()

['lowerLimit: -0.06, upperLimit: 0, nominalValue: 0.53, engineeringUnits: mm, cavity: 1']

## Seperating variable_attribute Feature:
* lowerLimit: -0.06499, upperLimit: 0.00499, nominalValue: 0.53, engineeringUnits: mm, cavity: 1

In [11]:
# We can use regex to map and seperate -> Remember for later use

In [12]:
def seperate_variable_attribute_feature(X:str):
    variable_attribute_list = X.split(",")
    variable_attribute_map = {}
    for i in variable_attribute_list:
        name , value = i.split(':')
        variable_attribute_map[name.strip()] = value.strip()
    return pd.Series(variable_attribute_map)

In [13]:
d2 = d2.join(d2.variable_attribute.apply(seperate_variable_attribute_feature))
d2.head()

device_name                 machine_definition        variable_name  \
2090171  de512stmp226  spark-machine-otto-vision-dynamic         stpStepFrIda   
2090157  de512stmp226  spark-machine-otto-vision-dynamic            teil2Wire   
2090158  de512stmp226  spark-machine-otto-vision-dynamic             teil2Iso   
2090159  de512stmp226  spark-machine-otto-vision-dynamic  iTeil2IsostempelDef   
2090160  de512stmp226  spark-machine-otto-vision-dynamic    folgefehlerTeil12   

                                                     value  \
2090171  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
2090157  1.9513,1.9592,1.9489,1.9397,1.962,1.9648,1.944...   
2090158  3.4863,3.4962,3.4957,3.4841,3.4901,3.4888,3.48...   
2090159  49.1378,48.7538,48.448,48.9035,48.923,48.8579,...   
2090160  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...   

                  timestamp  \
2090171 2025-12-05 08:15:13   
2090157 2025-12-05 08:15:13   
2090158 2025-12-05 08:15:13   
2090159 2025-12-05 08:15:13   
2090160 2025-12-05 08:15:13   

                                        variable_attribute lowerLimit  \
2090171  lowerLimit: -0.1, upperLimit: 0.1, nominalValu...       -0.1   
2090157  lowerLimit: -0.2, upperLimit: 0.2, nominalValu...       -0.2   
2090158  lowerLimit: -0.2, upperLimit: 0.2, nominalValu...       -0.2   
2090159  lowerLimit: -10, upperLimit: 10, nominalValue:...        -10   
2090160  lowerLimit: 0, upperLimit: 10, nominalValue: 0...          0   

        upperLimit nominalValue engineeringUnits cavity  
2090171        0.1            1               mm      1  
2090157        0.2            2               mm      1  
2090158        0.2          3.5               mm      1  
2090159         10           50               mm      1  
2090160         10            0               mm      1

In [14]:
# Now making these features float -> (Before they are object)
cols = ["lowerLimit", "upperLimit","nominalValue", "cavity"]
d2[cols] = d2[cols].astype(float)
d2.dtypes

device_name                   object
machine_definition            object
variable_name                 object
value                         object
timestamp             datetime64[ns]
variable_attribute            object
lowerLimit                   float64
upperLimit                   float64
nominalValue                 float64
engineeringUnits              object
cavity                       float64
dtype: object

## Remoiving the columns which have single value

In [15]:
col_val_map = {
    col_name: d2[col_name].unique() for col_name in d2.columns.tolist() if d2[col_name].dtype == 'object'
}

In [16]:
d2.cavity.unique()

array([1.])

In [17]:
col_val_map

{'device_name': array(['de512stmp226', 'de512stmp276', 'de512stmp268', 'de512stmp223',
        'de512stmp269', 'de512stmp278'], dtype=object),
 'machine_definition': array(['spark-machine-otto-vision-dynamic'], dtype=object),
 'variable_name': array(['stpStepFrIda', 'teil2Wire', 'teil2Iso', 'iTeil2IsostempelDef',
        'folgefehlerTeil12', 'teil1GapY', 'teil1GapX', 'iTeil1Rastfeder',
        'iTeil1Rastfederdifferenz', 'prfungTeil1KastenSpk', 'teil2GapY',
        'teil2GapX', 'iTeil2Rastfeder', 'iTeil2Rastfederdifferenz',
        'prfungTeil2BodySpk', 'gapLinks', 'testMittelwertTeil1GapX',
        'testMittelwertTeil1GapY', 'iTeil2Crimp',
        'testMittelwertTeil2GapX', 'iTeil2KastenHinten',
        'iTeil1IsostempelDef', 'iDeformationFeder',
        'iFreischnittLamelleUnten', 'iFreischnittLamelleOben',
        'iKastenVorne', 'iKastenHinten', 'wire', 'iso',
        'sp1DlErkennungErgebnis', 'sp2DlErkennungErgebnis',
        'l-tstelleBody', 'l-tstelleFeder', 'achtungTeil12Vertau

In [18]:
d2.columns

Index(['device_name', 'machine_definition', 'variable_name', 'value',
       'timestamp', 'variable_attribute', 'lowerLimit', 'upperLimit',
       'nominalValue', 'engineeringUnits', 'cavity'],
      dtype='object')

In [19]:
# Droping unwanted features
d2 = d2.drop(["machine_definition", "variable_attribute", "engineeringUnits", "cavity"], axis = 1)
d2.head()


device_name        variable_name  \
2090171  de512stmp226         stpStepFrIda   
2090157  de512stmp226            teil2Wire   
2090158  de512stmp226             teil2Iso   
2090159  de512stmp226  iTeil2IsostempelDef   
2090160  de512stmp226    folgefehlerTeil12   

                                                     value  \
2090171  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
2090157  1.9513,1.9592,1.9489,1.9397,1.962,1.9648,1.944...   
2090158  3.4863,3.4962,3.4957,3.4841,3.4901,3.4888,3.48...   
2090159  49.1378,48.7538,48.448,48.9035,48.923,48.8579,...   
2090160  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...   

                  timestamp  lowerLimit  upperLimit  nominalValue  
2090171 2025-12-05 08:15:13        -0.1         0.1           1.0  
2090157 2025-12-05 08:15:13        -0.2         0.2           2.0  
2090158 2025-12-05 08:15:13        -0.2         0.2           3.5  
2090159 2025-12-05 08:15:13       -10.0        10.0          50.0  
2090160 2025-12-05 08:15:13         0.0        10.0           0.0

## Converting the device name to name matching other datasets


In [20]:
def change_device_name(x):
    return f"S-{x[-3:]}"

d2['device_name'] = d2['device_name'].apply(change_device_name)
d2.head()

device_name        variable_name  \
2090171       S-226         stpStepFrIda   
2090157       S-226            teil2Wire   
2090158       S-226             teil2Iso   
2090159       S-226  iTeil2IsostempelDef   
2090160       S-226    folgefehlerTeil12   

                                                     value  \
2090171  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
2090157  1.9513,1.9592,1.9489,1.9397,1.962,1.9648,1.944...   
2090158  3.4863,3.4962,3.4957,3.4841,3.4901,3.4888,3.48...   
2090159  49.1378,48.7538,48.448,48.9035,48.923,48.8579,...   
2090160  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...   

                  timestamp  lowerLimit  upperLimit  nominalValue  
2090171 2025-12-05 08:15:13        -0.1         0.1           1.0  
2090157 2025-12-05 08:15:13        -0.2         0.2           2.0  
2090158 2025-12-05 08:15:13        -0.2         0.2           3.5  
2090159 2025-12-05 08:15:13       -10.0        10.0          50.0  
2090160 2025-12-05 08:15:13         0.0        10.0           0.0

In [21]:
d2.variable_name.unique()

array(['stpStepFrIda', 'teil2Wire', 'teil2Iso', 'iTeil2IsostempelDef',
       'folgefehlerTeil12', 'teil1GapY', 'teil1GapX', 'iTeil1Rastfeder',
       'iTeil1Rastfederdifferenz', 'prfungTeil1KastenSpk', 'teil2GapY',
       'teil2GapX', 'iTeil2Rastfeder', 'iTeil2Rastfederdifferenz',
       'prfungTeil2BodySpk', 'gapLinks', 'testMittelwertTeil1GapX',
       'testMittelwertTeil1GapY', 'iTeil2Crimp',
       'testMittelwertTeil2GapX', 'iTeil2KastenHinten',
       'iTeil1IsostempelDef', 'iDeformationFeder',
       'iFreischnittLamelleUnten', 'iFreischnittLamelleOben',
       'iKastenVorne', 'iKastenHinten', 'wire', 'iso',
       'sp1DlErkennungErgebnis', 'sp2DlErkennungErgebnis',
       'l-tstelleBody', 'l-tstelleFeder', 'achtungTeil12Vertauscht',
       'iTeil1KastenVorne', 'iTeil1KastenHinten', 'iTeil1Crimp',
       'teil1Wire', 'teil1Iso', 'iTeil2KastenVorne',
       'testMittelwertTeil2GapY', 'gapRechts', 'spur1Iso', 'spur1Gap',
       'spur2IKastenMitte', 'spur1Wire', 'spur1IKastenHinte

# Saving the current data

In [24]:
d2.to_parquet(
    r"E:\Learning\TEAI Cup\Data\Parquet Data\d2_1_general_cleaning.parquet", index=False
)

In [23]:
d2

device_name        variable_name  \
2090171       S-226         stpStepFrIda   
2090157       S-226            teil2Wire   
2090158       S-226             teil2Iso   
2090159       S-226  iTeil2IsostempelDef   
2090160       S-226    folgefehlerTeil12   
...             ...                  ...   
2089377       S-226            teil1GapY   
2089378       S-226            teil1GapX   
2089379       S-226      iTeil1Rastfeder   
2089373       S-226            teil2Wire   
2089371       S-226   iTeil2KastenHinten   

                                                     value  \
2090171  1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   
2090157  1.9513,1.9592,1.9489,1.9397,1.962,1.9648,1.944...   
2090158  3.4863,3.4962,3.4957,3.4841,3.4901,3.4888,3.48...   
2090159  49.1378,48.7538,48.448,48.9035,48.923,48.8579,...   
2090160  0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,...   
...                                                    ...   
2089377  0.5029,0.5018,0.4951,0.5008,0.5047,0.5,0.5043,...   
2089378  0.509,0.5071,0.506,0.5134,0.509,0.5134,0.5062,...   
2089379  44.3393,44.7688,46.0992,46.3617,46.1708,46.349...   
2089373  1.9311,1.9314,1.922,1.9411,1.9336,1.9271,1.935...   
2089371  1.8196,1.8211,1.8227,1.8282,1.8279,1.8197,1.81...   

                  timestamp  lowerLimit  upperLimit  nominalValue  
2090171 2025-12-05 08:15:13       -0.10        0.10          1.00  
2090157 2025-12-05 08:15:13       -0.20        0.20          2.00  
2090158 2025-12-05 08:15:13       -0.20        0.20          3.50  
2090159 2025-12-05 08:15:13      -10.00       10.00         50.00  
2090160 2025-12-05 08:15:13        0.00       10.00          0.00  
...                     ...         ...         ...           ...  
2089377 2025-12-10 11:08:06       -0.06        0.00          0.53  
2089378 2025-12-10 11:08:06       -0.06        0.00          0.53  
2089379 2025-12-10 11:08:06      -25.00       25.00         50.00  
2089373 2025-12-10 11:08:06       -0.20        0.20          2.00  
2089371 2025-12-10 11:08:06        0.00        0.15          1.75  

[2408396 rows x 7 columns]